proposals/0000-ghc2023.rst

```diff
@@ -0,0 +1,179 @@
+`GHC issue #19396 <https://gitlab.haskell.org/ghc/ghc/-/issues/19396>`_), and
+this can surprise users and cause backwards incompatibility.
+
+Since ``ExistentialQuantification`` allows defining types with contexts that
+include equality constraints, there is not really a principled distinction
+between ``ExistentialQuantification`` and ``GADTs``.  (While there is a
+syntactic distinction between GADT syntax and "traditional" datatype syntax,
+both forms are capable of expressing simple ADTs, existentially quantified
+types, and GADTs.)
+
+Possible ways to resolve this infelicity include:
+
+* Add ``GADTs`` and ``MonoLocalBinds`` to ``GHC2023``.  This makes it clear
+  that GADTs/existentials are a core part of the language, and makes the type
+  inference compromises necessary to accommodate them.  Migration advice for
+  ``GHC2023`` should make clear that type signatures may need to be added for
+  local bindings (or ``NoMonoLocalBinds`` specified explicitly).  Given that
+  ``MonoLocalBinds`` is a simpler design which can safely be extended with
+  ``GADTs``, it makes sense to have it be part of the base language; users can
+  then opt-in explicitly to ``NoMonoLocalBinds`` as an extension if required.
+
+* Add ``GADTs`` but not ``MonoLocalBinds``.  This is mostly consistent with
+  ``GHC2021``, but means that type inference for local bindings may not be
+  predictable when using GADTs.  Moreover, including an extension but not the
+  extensions it implies is itself confusing (as with
+  ``TypeOperators``/``ExplicitNamespaces`` in ``GHC2021``).
+
+* Remove ``ExistentialQuantification`` from ``GHC2023``.  This means
+  existentials/GADTs are clearly treated as an extension, albeit an extension
+  that makes type inference "worse". Users will need to understand the impact
+  of ``MonoLocalBinds`` only if they import a GADT or define one after enabling
+  ``GADTs`` explicitly. Migration advice for ``GHC2023`` should make clear that
+  users may need to enable ``GADTs`` explicitly (and possibly give type
+  signatures for local bindings or specify ``NoMonoLocalBinds``).
+
+
+Why add LambdaCase?
+-------------------
+
+The latest `State of Haskell 2021 Survey results
+<https://taylor.fausak.me/2022/11/18/haskell-survey-results/>`_ list
+`LambdaCase`_ as the top answer to “Which extension would you want to be on by
+default”. It also missed ``GHC2021`` by just two votes. There is a whole style
+of writing Haskell that makes extensive use of ``\case``. And (unlike the runner up in the survey, `OverloadedStrings_`), it only enables *new* syntax, i.e. it does not change existing code.
+
+A counter-agument to adding `LambdaCase` is that just extended the meaning of
+`LambdaCase`_ with ``\cases`` in `#302
+<https://github.com/ghc-proposals/ghc-proposals/pull/302>`_, and if one only
+wants to add extensions to ``GHC20xx`` that have been proven to be stable, then
+this one probably isn’t yet.
+
+
```

Comment:
```suggestion

Why add RoleAnnotations?
------------------------

Roles are essential part of modern GHC Haskell.

As `GeneralisedNewtypeDeriving` is in GHC2021 language, so should be `RoleAnnotations`. They are different sides of the same feature: without correct role annotations GND cannot be used safely.

Role annotations are required for correctly writing types with internal invariants like `Set` or "fast" implementations like `data Fin (n :: Nat) = UnsafeFin Int`.

````

Reply:
The latter is indeed a compelling point: `RoleAnnotations` aren’t really optional any more.